usecases/auth/authorization/rbac/manager.go

```diff
@@ -40,93 +51,135 @@ func (m *Manager) AddPolicies(policies []*Policy) error {
+			rolesPermissions[policy[0]] = append(rolesPermissions[policy[0]], permission(policy))
+		}
+	} else {
+		for _, name := range names {
+			polices, err := m.casbin.GetFilteredNamedPolicy("p", 0, name) // fmt.Sprintf("'%s' == p.sub", name)
+			if err != nil {
+				return nil, err
+			}
+			if len(polices) == 0 {
+				return nil, fmt.Errorf("%w: %s", authorization.ErrRoleNotFound, name)
+			}
+			for _, policy := range polices {
+				rolesPermissions[name] = append(rolesPermissions[name], permission(policy))
+			}
 		}
-		policies = append(policies, &Policy{Name: p[0], Resource: p[1], Verb: p[2], Domain: p[3]})
 	}
-	return policies, nil
+
+	for roleName, perms := range rolesPermissions {
+		roles = append(roles, &models.Role{
+			Name:        &roleName,
+			Permissions: perms,
+		})
+	}
+	return roles, nil
 }
 
-func (m *Manager) RemovePolicies(policies []*Policy) error {
-	for _, policy := range policies {
-		ok, err := m.casbin.RemovePolicy(policy.Name, policy.Resource, policy.Verb, policy.Domain)
+func (m *manager) DeleteRoles(roles ...string) error {
+	// TODO: block deleting built in roles
+	for _, role := range roles {
+		ok, err := m.casbin.RemoveFilteredNamedPolicy("p", 0, role)
 		if err != nil {
 			return err
 		}
 		if !ok {
-			return fmt.Errorf("failed to remove policy %v", policy)
+			return fmt.Errorf("failed to remove policy %v", role)
 		}
 	}
 	if err := m.casbin.SavePolicy(); err != nil {
 		return err
 	}
 
-	if err := m.casbin.InvalidateCache(); err != nil {
-		return err
-	}
-	return nil
+	return m.casbin.InvalidateCache()
 }
 
-func (m *Manager) AddRolesForUser(user string, roles []string) error {
+func (m *manager) AddRolesForUser(user string, roles []string) error {
+	// userName := fmt.Sprintf("%s%s", userPrefix, user)
 	for _, role := range roles {
+		// roleName := fmt.Sprintf("%s%s", rolePrefix, role)
 		if _, err := m.casbin.AddRoleForUser(user, role); err != nil {
 			return err
 		}
 	}
-	if err := m.casbin.SavePolicy(); err != nil {
-		return err
-	}
-	return nil
+
+	return m.casbin.SavePolicy()
 }
 
-func (m *Manager) GetRolesForUser(user string) ([]string, error) {
-	return m.casbin.GetRolesForUser(user)
+func (m *manager) GetRolesForUser(user string) ([]*models.Role, error) {
+	rolesNames, err := m.casbin.GetRolesForUser(user)
+	if err != nil {
+		return nil, err
+	}
+
+	return m.GetRoles(rolesNames...)
 }
 
-func (m *Manager) GetUsersForRole(role string) ([]string, error) {
+func (m *manager) GetUsersForRole(role string) ([]string, error) {
 	return m.casbin.GetUsersForRole(role)
 }
 
-func (m *Manager) DeleteRolesForUser(user string, role []string) error {
-	for _, role := range role {
+func (m *manager) RevokeRolesForUser(user string, roles ...string) error {
+	for _, role := range roles {
 		if _, err := m.casbin.DeleteRoleForUser(user, role); err != nil {
 			return err
 		}
 	}
 	if err := m.casbin.SavePolicy(); err != nil {
 		return err
 	}
-	return nil
+	return m.casbin.InvalidateCache()
```

Comment:
general question: When do we need to to invalidate this and when do we not? Only after deletes?

Reply:
deletes or adds any changes in the roles